storage/rocksdb/ha_rocksdb.cc

```diff
@@ -4164,6 +4196,52 @@ bool ha_rocksdb::init_with_fields() {
   DBUG_RETURN(false); /* Ok */
 }
 
+
+/*
+  If the key is a TTL key, we may need to filter it out.
+
+  The purpose of read filtering for tables with TTL is to ensure that
+  during a transaction a key which has expired already but not removed by
+  compaction yet is not returned to the user.
+
+  Without this the user might be hit with problems such as disappearing
+  rows within a transaction, etc, because the compaction filter ignores
+  snapshots when filtering keys.
+*/
+int ha_rocksdb::should_hide_ttl_rec(const rocksdb::Slice &ttl_rec_val,
+                                    const int64_t curr_ts) const {
+  DBUG_ASSERT(m_pk_descr != nullptr);
+  DBUG_ASSERT(m_pk_descr->has_ttl());
+
+  if (!rdb_is_ttl_read_filtering_enabled() || !rdb_is_ttl_enabled()) {
+    return false;
+  }
+
+  Rdb_string_reader reader(&ttl_rec_val);
+
+  uint64 ts;
+  if (reader.read_uint64(&ts)) {
+    /*
+      This condition should never be reached since all TTL records have an
+      8 byte ttl field in front. Don't filter the record out, and log an error.
+    */
+    std::string buf;
+    buf = rdb_hexdump(ttl_rec_val.data(), ttl_rec_val.size(),
+                      RDB_MAX_HEXDUMP_LEN);
+    const GL_INDEX_ID gl_index_id = m_pk_descr->get_gl_index_id();
+    // NO_LINT_DEBUG
+    sql_print_error("Decoding ttl from PK value failed, "
+                    "for index (%u,%u), val: %s",
+                    gl_index_id.cf_id, gl_index_id.index_id, buf.c_str());
+    DBUG_ASSERT(0);
+    return false;
+  }
+
+  /* Hide record if it has expired before the current snapshot time. */
+  uint64 time_diff = std::difftime(curr_ts ? curr_ts : std::time(nullptr), ts);
```

Comment:
If curr_ts is 0, then the time comparison constantly changes relative to the current time. It could be possible for some rows to disappear (i.e. secondary key passes, but the next read to the primary key returns nothing because it expired). Can you somehow guarantee the there is always a non-zero time passed in that is preserved for the duration of the query?